workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowLifecycleOwner.kt

```diff
@@ -0,0 +1,180 @@
+package com.squareup.workflow1.ui
+
+import android.view.View
+import android.view.View.OnAttachStateChangeListener
+import androidx.annotation.VisibleForTesting
+import androidx.annotation.VisibleForTesting.PRIVATE
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.Event
+import androidx.lifecycle.Lifecycle.State.DESTROYED
+import androidx.lifecycle.Lifecycle.State.RESUMED
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.LifecycleRegistry
+import androidx.lifecycle.LifecycleRegistry.createUnsafe
+import androidx.lifecycle.ViewTreeLifecycleOwner
+
+/**
+ * An extension of [LifecycleOwner] that is always owned by a [View], is logically a child lifecycle
+ * of the next-nearest [ViewTreeLifecycleOwner] above it (it mirrors its parent's lifecycle until
+ * it's destroyed), and can be [asked to destroy][destroyOnDetach] itself early.
+ *
+ * This type is meant to help integrate with [ViewTreeLifecycleOwner] by allowing the creation of a
+ * tree of [LifecycleOwner]s that mirrors the view tree.
+ *
+ * Custom container views that use [ViewRegistry.buildView] to create their children _must_ ensure
+ * they call [destroyOnDetach] on the outgoing view before they replace children with new views.
+ * If this is not done, then certain processes that are started by that view's subtree may continue
+ * to run long after the view is been detached, and memory and other resources may be leaked.
+ *
+ * Set a [WorkflowLifecycleOwner] on a view by calling [installOn], and read it back using [get].
+ */
+@WorkflowUiExperimentalApi
+public interface WorkflowLifecycleOwner : LifecycleOwner {
+
+  /**
+   * Flags this [lifecycle][Lifecycle] to be set to [DESTROYED] as soon as the owning view is
+   * [detached][View.onDetachedFromWindow].
+   */
+  public fun destroyOnDetach()
+
+  public companion object {
+    /**
+     * Creates a new [WorkflowLifecycleOwner] and sets it on [view]'s tags so it can be later
+     * retrieved with [get].
+     *
+     * It's very important that, once this function is called with a given view, that view gets
+     * attached at least once to ensure that the lifecycle eventually gets destroyed (either by
+     * being destroyed explicitly via [destroyOnDetach] or because its parent is destroyed).
+     * If this is not done, any observers registered with the [Lifecycle] may be leaked as they will
+     * never see the destroy event.
```

Comment:
I don't understand why this is necessary since we are by default hanging off the parent lifecycle anyway? if this view never got attached couldn't we make sure any LIfecycle observers got our destroyed event ?

Reply:
We don't get the parent lifecycle until the view is attached though. Although this is slightly incorrect, as it would also be ok to never attach the view but just call `destroyOnDetach()` directly – if the view is detached when this is called, the lifecycle will be destroyed immediately (TODO need docs here and on `destroyOnDetach`).

Reply:
Docs added.